### DISCUSSION TITLE
actisense-serial socket server?

### DISCUSSION THREAD

**Discussion by mglonnro** [canboat/canboat#382] [Q&A] [answered]
Cheers,

Is there some established way to make a bridge between the actisense-serial's serial in/out and a TCP socket server (ie actisense-serial is running as a daemon and talks to the USB device, and other applications and read/write to the stream over TCP)? Or some way to do it without actisense-serial? 

I'm thinking about adding some code to actisense-serial to support this, but I suspect someone has solved this already. 

Kind regards!

**Comment by sbender9** [comment]
https://github.com/SignalK/signalk-server

**Reply by mglonnro** [reply]
Thank you! Does it provide endpoints for the raw serial data as well?

**Reply by sbender9** [reply]
Not the raw data. But why would you want that? We are are dealing with that.

**Reply by mglonnro** [reply]
Hmm, for a few reasons (that may be wrong), but the one most relevant here is that I'm communicating with some devices using the NMEA 2000 lighting control standard, and I don't think this is implemented in SignalK?

**Reply by sbender9** [reply]
If that's not in canboat/js? Then contribute there.

**Comment by aldas** [comment]
There are examples for applications that spawn `actisense-serial` as subprocess and read stdio of it send output (over http in this case) somewhere else. https://github.com/iotfablab/n2kparser/blob/0453398034da5a11756d72f1b3c4459a13a5b39c/n2kparser/n2kparser.py#L82 

and you could add support to send stdio to that process as  `actisense-serial` support sending stdio to n2k bus. 
Here:
https://github.com/canboat/canboat/blob/a9739be2fd3ab4ba9cb22c80908499ce0ba98912/actisense-serial/actisense-serial.c#L363


but you probably can do same stuff with `nc` and/or `socat`  https://unix.stackexchange.com/questions/167204/redirect-stdin-and-stdout-to-ports but this is quite limited

**Reply by mglonnro** [reply]
Thank you!

**Comment by keesverruijt** [comment, marked as answer]
Take a look at `n2kd_monitor`, I think either it already does what you need or it is simple to adapt for your purposes. You could replace the `analyzer` and `n2kd` programs in the pipeline that it builds with `socat` or `nc`. You could also extend the `iptee` program that comes with canboat so that it also reads from any TCP clients. 

https://github.com/canboat/canboat/wiki/n2kd_monitor

Note that the standard `n2kd_monitor` pipeline already allows writing of _any_ N2K message, regardless of whether `analyzer` understands your (maybe proprietary) messages. I can see how it would be useful if `analyzer` and `n2kd` would pass on messages that they do not understand in some raw unexpanded format.  Currently an error will be logged instead.

**Comment by keesverruijt** [comment]
And as @sbender9 says -- if you are using our open source tools, please contribute back and add missing PGNs.

I have added (very) preliminary support to CANBoat for the PGNs added in NMEA 2000 v3.002 -- electrical motors, elevator cars (!!!) and lighting.

If you have a log file containing the lighting PGNs can you please share this? And if you have reverse engineered the lighting data lengths etc. it would be nice if you shared that -- either in text form or a pull request.

**Reply by mglonnro** [reply]
Yes, absolutely! I don't have any logs yet but will at some point, so I'll share those then.

**Reply by mglonnro** [reply]
One more thing: if I try to add missing PGNs to analyzer, what is the master file I should understand & use? Most of this light stuff goes through Request Group Function/Command Group Function PGN 126208.

**Reply by keesverruijt** [reply]
The definitions are in pgn.h. Search for “lighting”, the base pgns have been added, but the field lengths and resolutions are just wild guesses right now. 

Do you need help with how the group functions work?

**Reply by mglonnro** [reply]
Thank you! I think I have a basic understanding of the group functions (at least I'm able to create actisense-serial input that does what I want with the lighting controller). 

I started experimenting with pgn.h, and that too seems reasonably straightforward. 

However, after updating the first field lengths (130330), I did get something almost reasonable out of analyzer, but the light controller seems to broadcast these 130330s in 8-byte packets? analyzer doesn't make sense of that since it decodes them individually. Here's a response to a system on command: 

```
2023-03-07T13:45:28.994Z,3,126208,92,64,7,02,1a,fd,01,00,01,f0
2023-03-07T13:45:28.994Z,7,130330,92,255,8,20,24,21,18,01,53,43,2d
2023-03-07T13:45:28.994Z,7,130330,92,255,8,21,4c,69,67,68,74,69,6e
2023-03-07T13:45:28.995Z,7,130330,92,255,8,22,67,20,43,6f,6e,74,72
2023-03-07T13:45:28.997Z,7,130330,92,255,8,23,6f,6c,6c,65,72,28,06
2023-03-07T13:45:28.997Z,7,130330,92,255,8,24,06,28,14,07,01,ff,ff
2023-03-07T13:45:28.997Z,7,130330,92,255,8,25,ff,ff,ff,ff,ff,ff,ff
2023-03-07T13:45:28.997Z,7,130330,92,255,8,40,24,21,18,01,53,43,2d
2023-03-07T13:45:28.997Z,7,130330,92,255,8,41,4c,69,67,68,74,69,6e
2023-03-07T13:45:28.998Z,7,130330,92,255,8,42,67,20,43,6f,6e,74,72
2023-03-07T13:45:28.998Z,7,130330,92,255,8,43,6f,6c,6c,65,72,28,06
2023-03-07T13:45:28.998Z,7,130330,92,255,8,44,06,28,14,07,01,ff,ff
2023-03-07T13:45:29.000Z,7,130330,92,255,8,45,ff,ff,ff,ff,ff,ff,ff
```

**Reply by mglonnro** [reply]
Oh, ok. Actually, when I run analyzer just on those lines (test2.txt extracted from the whole log file), it _does_ work:

```
2023-03-07T13:45:24.696Z 7  92 255 130330 Lighting System Settings:  Global Enable = 0; Default Settings/Commands = 0; Reserved = 20; Name of the lighting controller = SC-Lighting Controller; Max Scenes = 40; Max Scene Configuration Count = 6; Max Zones = 6; Max Color Sequences = 40; Max Color Sequence Color Count = 20; Number of Programs = 7; Controller Capabilities = 1; Identify Device = Unknown
2023-03-07T13:45:24.699Z 7  92 255 130330 Lighting System Settings:  Global Enable = 0; Default Settings/Commands = 0; Reserved = 20; Name of the lighting controller = SC-Lighting Controller; Max Scenes = 40; Max Scene Configuration Count = 6; Max Zones = 6; Max Color Sequences = 40; Max Color Sequence Color Count = 20; Number of Programs = 7; Controller Capabilities = 1; Identify Device = Unknown
ERROR 2023-03-07T14:52:37.495Z [analyzer] Received incomplete fast packet PGN 126208 from source 92
2023-03-07T13:45:28.997Z 7  92 255 130330 Lighting System Settings:  Global Enable = 1; Default Settings/Commands = 0; Reserved = 20; Name of the lighting controller = SC-Lighting Controller; Max Scenes = 40; Max Scene Configuration Count = 6; Max Zones = 6; Max Color Sequences = 40; Max Color Sequence Color Count = 20; Number of Programs = 7; Controller Capabilities = 1; Identify Device = Unknown
2023-03-07T13:45:29.000Z 7  92 255 130330 Lighting System Settings:  Global Enable = 1; Default Settings/Commands = 0; Reserved = 20; Name of the lighting controller = SC-Lighting Controller; Max Scenes = 40; Max Scene Configuration Count = 6; Max Zones = 6; Max Color Sequences = 40; Max Color Sequence Color Count = 20; Number of Programs = 7; Controller Capabilities = 1; Identify Device = Unknown
```

But when it sees the whole log file (test.txt), it fails: 
```
2023-03-07T13:45:28.994Z 3  92  64 126208 NMEA - Acknowledge group function:  Function Code = 2; PGN = 130330; PGN error code = Acknowledge; Transmission interval/Priority error code = Acknowledge; Number of Parameters = 1; Parameter 1 = Acknowledge
ERROR 2023-03-07T14:53:49.897Z [analyzer] Unhandled string type 33 in PGN
2023-03-07T13:45:28.994Z 7  92 255 130330 Lighting System Settings:  Global Enable = 0; Default Settings/Commands = 0; Reserved = 20
ERROR 2023-03-07T14:53:49.897Z [analyzer] PGN 130330 analysis error
ERROR 2023-03-07T14:53:49.897Z [analyzer] Unhandled string type 105 in PGN
2023-03-07T13:45:28.994Z 7  92 255 130330 Lighting System Settings:  Global Enable = 1; Default Settings/Commands = 0; Reserved = 20
ERROR 2023-03-07T14:53:49.898Z [analyzer] PGN 130330 analysis error
ERROR 2023-03-07T14:53:49.898Z [analyzer] Unhandled string type 32 in PGN
2023-03-07T13:45:28.995Z 7  92 255 130330 Lighting System Settings:  Global Enable = 2; Default Settings/Commands = 0; Reserved = 20
ERROR 2023-03-07T14:53:49.898Z [analyzer] PGN 130330 analysis error
ERROR 2023-03-07T14:53:49.899Z [analyzer] Unhandled string type 108 in PGN
2023-03-07T13:45:28.997Z 7  92 255 130330 Lighting System Settings:  Global Enable = Unknown; Default Settings/Commands = 0; Reserved = 20
ERROR 2023-03-07T14:53:49.899Z [analyzer] PGN 130330 analysis error
ERROR 2023-03-07T14:53:49.899Z [analyzer] Unhandled string type 40 in PGN
```
[test.txt](https://github.com/canboat/canboat/files/10910725/test.txt)
[test2.txt](https://github.com/canboat/canboat/files/10910726/test2.txt)

**Comment by keesverruijt** [comment]
I think the answer is that this is a "bug" in the Actisense NGT-1 firmware. It assumes that PGNs that it doesn't know in this range are "single", so the NGT sends a message per packet of the fast framed PGN. This needs to be addressed by Actisense. I checked but the current release for the NGT-1 is already quite old so they need to deliver an update.

The reason that `analyzer` behaves the way it does is that as soon as it sees a coalesced record containing all bytes of any Fast packet (> 8 bytes) is that it switches from `RAWFORMAT_PLAIN` into `RAWFORMAT_FAST` mode and assumes that all records sent to it will be sent the same way. This stems from the days where I did not have a clear idea of which packets were fast PGNS and which were not. 

It is not impossible to fix this in the analyzer, where it recognises that it is being sent _so_ little information for a PGN that it should assume that it needs to do fast packet re-assembly for this PGN even when it shouldn't for others. This is a little peculiar and might need a new mode, but it is not impossible.

For now you could get away with using a UNIX pipeline split, like this:

    actisense-serial  ---> tee ---> grep ,130330, -----> analyzer --\
                            |                                        \
                            \-----> grep -v ,130330, ---> analyzer ---> 
                                           
Here is an example how to do this using BASH process substitution:

    #!/usr/bin/env bash
    exec 4>&1
    {
       head samples/strings.raw
       cat  samples/pgn130330.raw
       #or in your case, simply actisense-serial
    } | tee >( grep ,130330, | ./rel/darwin-arm64/analyzer -q -json >&4 ) |
         grep -v ,130330, | ./rel/darwin-arm64/analyzer -q -json

**Reply by mglonnro** [reply]
Cool! And I'm actually using the Yacht Devices YDNU-02 so at least that one behaves as you describe.

**Comment by keesverruijt** [comment]
There is now a new `analyzer -format PLAIN_OR_FAST` option that should fix your problem.

You may want to contact Yacht Devices and ask them for NMEA 2000 v3 support.

**Reply by mglonnro** [reply]
Wow! It works beautifully!

```
{"version":"4.10.1","units":"std","showLookupValues":false}
{"timestamp":"2023-03-07T18:34:00.133Z","prio":7,"src":92,"dst":255,"pgn":126993,"description":"Heartbeat","fields":{"Data transmit offset":"00:01:00","Sequence Counter":189,"Controller 1 State":"Error Active","Controller 2 State":"Error Active","Equipment Status":"Operational"}}
{"timestamp":"2023-03-07T18:34:05.884Z","prio":7,"src":92,"dst":255,"pgn":130561,"description":"Lighting Zone","fields":{"Zone Index":2,"Color Temperature":0,"Intensity":192,"Program Color Sequence Index":0,"Program Intensity":200,"Program Rate":200,"Program Color Sequence":200,"Zone Enabled":"Off","Reserved":"00"}}
ERROR 2023-03-07T18:34:37.968Z [analyzer] Received incomplete fast packet PGN 126208 from source 92
{"timestamp":"2023-03-07T18:34:37.969Z","prio":7,"src":92,"dst":255,"pgn":130561,"description":"Lighting Zone","fields":{"Zone Index":2,"Color Temperature":0,"Intensity":192,"Program Color Sequence Index":0,"Program Intensity":200,"Program Rate":200,"Program Color Sequence":200,"Zone Enabled":"On","Reserved":"00"}}
{"timestamp":"2023-03-07T18:35:02.619Z","prio":7,"src":92,"dst":255,"pgn":126993,"description":"Heartbeat","fields":{"Data transmit offset":"00:01:00","Sequence Counter":190,"Controller 1 State":"Error Active","Controller 2 State":"Error Active","Equipment Status":"Operational"}}
{"timestamp":"2023-03-07T18:36:05.087Z","prio":7,"src":92,"dst":255,"pgn":126993,"description":"Heartbeat","fields":{"Data transmit offset":"00:01:00","Sequence Counter":191,"Controller 1 State":"Error Active","Controller 2 State":"Error Active","Equipment Status":"Operational"}}
{"timestamp":"2023-03-07T18:36:19.740Z","prio":6,"src":92,"dst":255,"pgn":60928,"description":"ISO Address Claim","fields":{"Unique Number":1284905,"Manufacturer Code":1084,"Device Instance Lower":0,"Device Instance Upper":0,"Device Function":130,"System Instance":0,"Industry Group":"Marine","Arbitrary address capable":1}}
{"timestamp":"2023-03-07T18:36:20.023Z","prio":7,"src":92,"dst":255,"pgn":130330,"description":"Lighting System Settings","fields":{"Global Enable":0,"Default Settings/Command":"Powering Up","Reserved":"20","Name of the lighting controller":"SC-Lighting Controller","Max Scenes":40,"Max Scene Configuration Count":6,"Max Zones":6,"Max Color Sequences":40,"Max Color Sequence Color Count":20,"Number of Programs":7,"Controller Capabilities":1}}
{"timestamp":"2023-03-07T18:36:20.030Z","prio":7,"src":92,"dst":255,"pgn":130561,"description":"Lighting Zone","fields":{"Zone Index":0,"Zone Name":"Zone 1","Red Component":16,"Green Component":16,"Blue Component":0,"Color Temperature":0,"Intensity":0,"Program ID":1,"Program Color Sequence Index":0,"Program Intensity":0,"Program Rate":200,"Program Color Sequence":200,"Zone Enabled":"Off","Reserved":"00"}}
{"timestamp":"2023-03-07T18:36:20.044Z","prio":7,"src":92,"dst":255,"pgn":130561,"description":"Lighting Zone","fields":{"Zone Index":1,"Zone Name":"Zone 2","Red Component":16,"Green Component":16,"Blue Component":0,"Color Temperature":0,"Intensity":0,"Program ID":1,"Program Color Sequence Index":0,"Program Intensity":0,"Program Rate":200,"Program Color Sequence":200,"Zone Enabled":"Off","Reserved":"00"}}
{"timestamp":"2023-03-07T18:36:20.059Z","prio":7,"src":92,"dst":255,"pgn":130561,"description":"Lighting Zone","fields":{"Zone Index":2,"Color Temperature":0,"Intensity":192,"Program Color Sequence Index":0,"Program Intensity":200,"Program Rate":200,"Program Color Sequence":200,"Zone Enabled":"Off","Reserved":"00"}}
{"timestamp":"2023-03-07T18:36:20.075Z","prio":7,"src":92,"dst":255,"pgn":130561,"description":"Lighting Zone","fields":{"Zone Index":3,"Zone Name":"Zone 4","Red Component":16,"Green Component":16,"Blue Component":0,"Color Temperature":0,"Intensity":0,"Program ID":1,"Program Color Sequence Index":0,"Program Intensity":0,"Program Rate":200,"Program Color Sequence":200,"Zone Enabled":"Off","Reserved":"00"}}
{"timestamp":"2023-03-07T18:36:20.092Z","prio":7,"src":92,"dst":255,"pgn":130561,"description":"Lighting Zone","fields":{"Zone Index":4,"Zone Name":"Zone 5","Red Component":0,"Green Component":0,"Blue Component":0,"Color Temperature":0,"Intensity":0,"Program ID":1,"Program Color Sequence Index":0,"Program Intensity":0,"Program Rate":200,"Program Color Sequence":200,"Zone Enabled":"Off","Reserved":"00"}}
{"timestamp":"2023-03-07T18:36:20.106Z","prio":7,"src":92,"dst":255,"pgn":130561,"description":"Lighting Zone","fields":{"Zone Index":5,"Zone Name":"Zone 6","Red Component":0,"Green Component":0,"Blue Component":0,"Color Temperature":0,"Intensity":0,"Program ID":1,"Program Color Sequence Index":0,"Program Intensity":0,"Program Rate":200,"Program Color Sequence":200,"Zone Enabled":"Off","Reserved":"00"}}
{"timestamp":"2023-03-07T18:36:20.378Z","prio":7,"src":92,"dst":255,"pgn":130563,"description":"Lighting Device","fields":{"Device ID":1457204591,"Device Capabilities":1,"Color Capabilities":7,"Zone Index":2,"Name of Lighting Device":"SBL [137]","Status":1,"Red Component":0,"Green Component":0,"Blue Component":0,"Color Temperature":0,"Intensity":0,"Program ID":0,"Program Color Sequence Index":0,"Program Intensity":0,"Program Rate":0,"Program Color Sequence Rate":0,"Enabled":"On","Reserved":"00"}}
{"timestamp":"2023-03-07T18:36:27.182Z","prio":7,"src":92,"dst":255,"pgn":130563,"description":"Lighting Device","fields":{"Device ID":1457204591,"Device Capabilities":1,"Color Capabilities":7,"Zone Index":2,"Name of Lighting Device":"SBL [137]","Status":0,"Red Component":0,"Green Component":0,"Blue Component":0,"Color Temperature":0,"Intensity":0,"Program ID":0,"Program Color Sequence Index":0,"Program Intensity":0,"Program Rate":0,"Program Color Sequence Rate":0,"Enabled":"On","Reserved":"00"}}
{"timestamp":"2023-03-07T18:37:22.194Z","prio":7,"src":92,"dst":255,"pgn":126993,"description":"Heartbeat","fields":{"Data transmit offset":"00:01:00","Sequence Counter":0,"Controller 1 State":"Error Active","Controller 2 State":"Error Active","Equipment Status":"Operational"}}
```
